reviewing/dependencies-hierarchy/test/getTree.test.ts

```diff
@@ -147,4 +147,334 @@ describe('getTree', () => {
       ])
     })
   })
+
+  // This group of tests attempts to check that package tree caching still
+  // respects max depth. See https://github.com/pnpm/pnpm/issues/4814
+  //
+  // This doesn't test the cache directly, but sets up situations that would
+  // result in incorrect output if the cache was used when it's not supposed to.
+  describe('prints at expected depth for cache regression testing cases', () => {
+    const commonMockGetTreeArgs = {
+      modulesDir: '',
+      includeOptionalDependencies: false,
+      skipped: new Set<string>(),
+      registries: {
+        default: 'mock-registry-for-testing.example',
+      },
+    }
+
+    test('revisiting package at lower depth prints dependenices not previously printed', () => {
+      // This tests the "glob" npm package on a subset of its dependency tree.
+      // Requested depth (max depth - current depth) shown in square brackets.
+      //
+      // root
+      // └─┬ glob [2]
+      //   ├─┬ inflight [1]
+      //   │ └── once [0]    <-- 1st time seen. No dependencies of "once" printed due to max depth.
+      //   └─┬ once [1]      <-- 2nd time seen, but at different depth. The "wrappy" dependency below should be printed.
+      //     └── wrappy [0]
+      //
+      const version = '1.0.0'
+      const currentPackages = generateMockCurrentPackages(version, {
+        root: ['glob'],
+        glob: ['inflight', 'once'],
+        inflight: ['once'],
+        once: ['wrappy'],
+      })
+      const rootDepPath = refToRelativeOrThrow(version, 'root')
+
+      const result = getTree({
+        ...commonMockGetTreeArgs,
+        currentDepth: 0,
+        maxDepth: 2,
+        currentPackages,
+        wantedPackages: currentPackages,
+      }, [rootDepPath], rootDepPath)
+
+      expect(normalizePackageNodeForTesting(result)).toEqual([
+        // depth 0
+        expect.objectContaining({
+          alias: 'glob',
+          dependencies: expect.arrayContaining([
+
+            // depth 1
+            expect.objectContaining({
+              alias: 'inflight',
+              dependencies: expect.arrayContaining([
+                // depth 2
+                expect.objectContaining({
+                  // The "once" package is first seen here at depth 2.
+                  alias: 'once',
+                }),
+              ]),
+            }),
+
+            // depth 1
+            expect.objectContaining({
+              alias: 'once',
+              dependencies: [
+                // The "once" package is seen again at depth 1. The "once"
+                // package contains a "wrappy" package that should be listed.
+                expect.objectContaining({ alias: 'wrappy' }),
+              ],
+            }),
+          ]),
+        }),
+      ])
+    })
+
+    test('revisiting package at higher depth does not print extra dependenices', () => {
```

Comment:
This is a new test added in the latest iteration of the PR. It tests the inverse situation of the test above. This test also fails before the fixes in this PR.